simplify work with containers

shaforostoff opened the pull request:
also do not create temporary keys container,
iterate directly over map

shaforostoff commented on engine/src/qlcfixturemode.cpp:
```diff
@@ -264,11 +258,10 @@ quint32 QLCFixtureMode::channelActsOn(quint32 chIndex)
     if (actsOnIndex != QLCChannel::invalid())
         m_actsOnMap[chIndex] = actsOnIndex;
+    else
+        m_actsOnMap.remove(chIndex);
```
this reduces number of map iterations

shaforostoff commented on engine/src/qlcfixturemode.h:
```diff
@@ -320,7 +320,6 @@ class QLCFixtureMode
 
     /** Save a mode to an XML document */
     bool saveXML(QXmlStreamWriter *doc);
-    QHash<QLCChannel *, QLCChannel *> actsOnChannelsList() const;
```
there is no definition of this method in cpp file

shaforostoff commented on engine/src/fixture.cpp:
```diff
@@ -1428,7 +1425,6 @@ bool Fixture::saveXML(QXmlStreamWriter *doc) const
         doc->writeTextElement(KXMLFixtureForcedLTP, list);
     }
 
-    if (m_channelModifiers.isEmpty() == false)
```
no need to check if the map is empty. If it is empty, the loop below will have zero iterations, so de facto isEmpty check is being made by it.hasNext()

shaforostoff commented on engine/src/universe.cpp:
```diff
@@ -216,25 +216,18 @@ QSharedPointer<GenericFader> Universe::requestFader(Universe::FaderPriority prio
-        if (m_faders.isEmpty())
+        for (int i = m_faders.count() - 1; i >= 0; i--)
         {
-            m_faders.append(fader);
```
no need to this special case: If  m_faders.count()==0, then the for loop will have zero iterations, and insertPos will remain 0. insert(0, fader) in this case is an equivalent of both append and prepend, according to Qt documentation

mcallegari commented:
Hi, thanks for this but please don't use the "auto" type. It's C++11 and the QLC+ engine still preserve older GCCs compatibility.
This is not the case for the QMLUI code path.
When v5 will become the default, the engine can be ported to more modern C++ standard.
Thanks

shaforostoff commented:
I have pushed the change replacing auto with real types. BTW, it would be nice if macOS builds supported older macOS versions, ideally macOS 12 Monterey. Right now I have to start the app via commandline: /Applications/QLC+.app/Contents/MacOS/qlcplus

mcallegari commented on engine/src/qlcfixturemode.cpp:
```diff
@@ -64,13 +64,7 @@ QLCFixtureMode& QLCFixtureMode::operator=(const QLCFixtureMode& mode)
-            ait.next();
-            m_actsOnMap.insert(ait.key(), ait.value());
-        }
+        m_actsOnMap = mode.m_actsOnMap;
```
Are you sure this is a deep copy like the previous code?

shaforostoff replied:
QMap is implicitly shared. The underlying data will be deep-copied the next time either of the two is modified (strictly speaking, when non-const member is invoked on either of them).

So if you know you want to force deep copy at this point (instead of later when/if it will be actually needed), I can add m_actsOnMap.begin() call after this line. Since m_actsOnMap is non-const, a non-const begin() will be invoked which will in turn call detach() internally which will lead to a forced deep copy.

In general, I would like to convert all QMap and QHash uses in the app to std::flat_map after you allow engine code to use c++23 features. This will reduce memory fragmentation and hence its usage, and will speed up map/hash access as well. std::flat_map is not implicitly shared and a deep copy will happen already during the assignment.

shaforostoff commented:
I have rebased and added simplification in In Function::requestAttributeOverride which reduces complexity from O(n*log_n) to O(n)

Before my change it was doing new map lookup for each key of this map

mcallegari commented:
> Also, may I ask why c++99 standard compatibility has to be preserved, given that the application uses Qt5? Which QLC+ target environments are not supporting at least c++17?

QLC+ is built on a wide variety of platforms and OSes. Starting from OBS, where old OSes has been kept for a long time, the Raspberry Pi, macOS, etc..
As of today, it should be safe to use C++11 and perhaps C++17 features.
However, PLEASE don't use the `auto` type! I hate `auto` and it feels so Java-ish.
Let's use it only where it makes life easier like replacing something like `QMap<QVariant, QPair<QString, QByteArray>>`
Also, I believe `auto` makes quick code checks from the GitHub web pages (I do this a lot) harder as you always have to remember the type of a variable or jump to the header file to find out.
Agree? :slightly_smiling_face:

mcallegari commented:
> I don't know much about OBS - could you please provide me with some links?

OBS stands for OpenSUSE Build Service.
https://build.opensuse.org/project/show/home:mcallegari79

It is a nice way to automate builds for several Linux distro at once.

shaforostoff commented:
I have split the most simple changes into https://github.com/mcallegari/qlcplus/pull/1726

Now this PR is smaller in size

mcallegari commented:
Thanks :+1:
